server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1017,6 +1017,16 @@ public boolean refreshNeeded() {
     @Nullable
     public abstract RefreshResult refresh(String source) throws EngineException;
 
+    /**
+     * An async variant of {@link Engine#refresh(String)} that may apply some rate-limiting.
+     */
+    public void externalRefresh(String source, ActionListener<Engine.RefreshResult> listener) {
```

Comment:
If I'm understanding how this is going to be used correctly, something like `throttledRefresh` or `rateLimitedRefresh` seems also like a good name.